miniscope_io/models/pipeline.py

```diff
@@ -0,0 +1,250 @@
+    config: NodeConfig
+
+    input_type: ClassVar[type[T]]
+    inputs: dict[str, Union["Source", "ProcessingNode"]] = Field(default_factory=dict)
+    output_type: ClassVar[type[U]]
+    outputs: dict[str, Union["Sink", "ProcessingNode"]] = Field(default_factory=dict)
+
+    @abstractmethod
+    def start(self) -> None:
+        """
+        Start producing, processing, or receiving data
+        """
+
+    @abstractmethod
+    def stop(self) -> None:
+        """
+        Stop producing, processing, or receiving data
+        """
+
+    @classmethod
+    def from_config(cls, config: NodeConfig) -> Self:
+        """
+        Create a node from its config
+        """
+        return cls(id=config.id, config=config)
+
+    @classmethod
+    @final
+    def node_types(cls) -> dict[str, type["Node"]]:
+        """
+        Map of all imported :attr:`.Node.type_` names to node classes
+        """
+        node_types = {}
+        to_check = cls.__subclasses__()
+        while to_check:
+            node = to_check.pop()
+            if node.type_ in node_types:
+                raise ValueError(
+                    f"Repeated node type_ identifier: {node.type_}, found in:\n"
+                    f"- {node_types[node.type_]}\n- {node}"
+                )
+            node_types[node.type_] = node
+            to_check.extend(node.__subclasses__())
+        return node_types
+
+
+class Source(Node, Generic[T, U]):
+    """A source of data in a processing pipeline"""
+
+    inputs: Final[None] = None
+    input_type: ClassVar[None] = None
+
+    @abstractmethod
+    def process(self) -> U:
+        """
+        Process some data, returning an output.
+
+
+        .. note::
+
+            The `process` method should not directly call or pass
+            data to subscribed output nodes, but instead return the output
+            and allow a containing pipeline class to handle dispatching data.
+
+        """
+
+
+class Sink(Node, Generic[T, U]):
+    """A sink of data in a processing pipeline"""
+
+    output_type: ClassVar[None] = None
+    outputs: Final[None] = None
+
+    @abstractmethod
+    def process(self, data: T) -> None:
+        """
+        Process some incoming data, returning None
+
+        .. note::
+
+            The `process` method should not directly be called or passed data,
+            but instead should be called by a containing pipeline class.
+
+        """
+
+
+class ProcessingNode(Node, Generic[T, U]):
+    """
+    An intermediate processing node that transforms some input to output
+    """
+
+    @abstractmethod
+    def process(self, data: T) -> U:
+        """
+        Process some incoming data, yielding a transformed output
+
+        .. note::
+
+            The `process` method should not directly call or be called by
+            output or input nodes, but instead return the output
+            and allow a containing pipeline class to handle dispatching data.
+
+        """
```

Comment:
Just wondering about minor stuff:
- I think these are going to run asynchronously, but is my understanding correct that these processes will typically have queues on the input side? I wondered how data flows, especially when there are parallel connections, and if this `T` can only be an instance of data or can also be something like the queue identifier. 
- Will these `T` and `U`s be defined so users can know what makes sense to connect? It probably depends on the intended scale, but I was wondering how people will define config (will there be type hint-ish things, will they find first out by the input validation presumably running for each process input, etc.).

Reply:
> I wondered how data flows, especially when there are parallel connections, and if this T can only be an instance of data or can also be something like the queue identifier.

The goal with making the processing functions into "pure" functions that don't handle things like queueing or getting data to/from other processing nodes is that we can then wrap them with various strategies for running them. So say we have one node `ProcessingTask`, we might imagine that having some wrapper for that like `MultiprocessRunner` like this

```python
class MultiprocessRunner(mp.Process):

    def __init__(self, node: Node):
        self.in_q = mp.Queue()
        self.out_q = mp.Queue()
        self.node = node

    def run():
        for item in iter(self.in_q, None):
            output = self.node.process(item)
            self.out_q.put(output)
```

But then we might also want to have just a purely synchronous runner that just passes input directly to the output. Or we might want a socket-based runner like this

```python
class SocketRunner():
    def __init__(self, node: Node):
        self.context = zmq.Context()
        self.socket = context.Socket(zmq.REP)
        self.socket.bind("ipc://*:5555")
        self.socket.on_recv(self.process)

    def self.process(msg):
        output = self.node.process(msg)
        self.socket.send(output)
```

and so on. So separating the queuing from the operation of the processing node is v much intentional and a strength: then we can handle developing the logic of passing queues separately from the actual logic of the processing stage.

> Will these T and Us be defined so users can know what makes sense to connect?

Yes! So that's the `input_type` and the `output_type` classvars. A given node will only be able to handle a specific type of input and a specific type of output, and so the `TypeVar`s `T` and `U` take their definition from the value of that parameter. That notation just says "the `process` method will return whatever `T` means in this context." so they are purposely opaque. A concrete child of those abstract classes would do something like

```python
class ParseHeader(ProcessingNode):
    input_type: ClassVar[type[np.ndarray]]
    output_type: FrameHeader
    
    # then implicitly, the signature becomes...
    def process(self, value: np.ndarray) -> FrameHeader: ...
```

They shouldn't need to be defined in config anywhere, the `TypeVars` are only used in `ClassVar`s or in method signatures here.

> InputType = TypeVar("InputType") to make it more readable?

This is mostly a convention, where they are purposely generic because they take on meaning from their use, so there isn't a technical reason that they can't have expressive names like that aside from a name like "`T`" is the conventional signal that something is a `TypeVar` - most of the time people shouldn't need to be interacting with them directly, since they're a generic part of the abstract parent classes that take their values from those given to the concrete child classes. but ya if you want to rename these there's not a strong reason why *not* aside from convention

Reply:
Thanks for the answers!
- What would you think about including the runner examples in the pipeline framework code? The Runner wrapper examples clarify stuff, at least for me, because the term pipeline makes me curious about how data is passed.
- For `TypeVar` convention either way works for me. If it's common I should probably get used to it.

Reply:
> What would you think about including the runner examples in the pipeline framework code?

yes! I am planning on working on those as i convert the existing camera classes, and they will probably evolve as i do so, so i will definitely add docs as i write them.